system/lib/libc/musl/src/signal/setitimer.c

```diff
@@ -37,11 +38,25 @@ void _emscripten_timeout(int which, double now)
 		signum = SIGPROF;
 	else if (which == ITIMER_VIRTUAL)
 		signum = SIGVTALRM;
-	int next_timeout = current_intervals_ms[which];
-	if (next_timeout)
-		current_timeout_ms[which] = now + next_timeout;
-	else
+	double next_timeout = 0.0;
+	if (current_intervals_ms[which]) {
+		// If time went backwards, schedule the next timer as if it didn't.
+		now = __builtin_wasm_max_f64(now, current_timeout_ms[which]);
```

Comment:
Do we really need to resort to a wasm specific builtin here?

Reply:
The reason I did this is basically just for code size. Using the MAX macro (or fmax) ends up being noticeably more than one instruction because of differences in the way NaNs are handled. I was initially on the fence about now much we cared about this rare corner case but I felt better handling it with this very cheap solution.

Reply:
Also, is there no general clang/llvm builtin that isn't wasm specific?

Reply:
Yeah, LLVM has [llvm.maxnum](https://llvm.org/docs/LangRef.html#llvm-maxnum-intrinsic) and `llvm.maximum` and they differ in the NaN handling. IIRC `MAX` and libc's `fmax` boils down to one of those (maxnum IIRC) but wasm's behavior matches `llvm.maximum`. I'm not sure whether there's a clang builtin for llvm.maxiumum but my thought was if we're already going to be using clang-specific builtins, we might as well use the one that corresponds exactly to what we want.

Reply:
BTW, this codesize of this function is not super critical since it won't get linked into programs that don't use timers (we reference it weakly when we use it elsewhere).   Just FYI